internal/metrics/circuitbreaker.go

```diff
@@ -0,0 +1,171 @@
+	IncClose(controller string)
+	IncEvent(controller, result string)
+}
+
+type cbMetrics struct {
+	opens  *prometheus.CounterVec
+	closes *prometheus.CounterVec
+	events *prometheus.CounterVec
+
+	controllers sync.Map // map[string]*controllerCounters
+}
+
+type controllerCounters struct {
+	opens  prometheus.Counter
+	closes prometheus.Counter
+
+	mu     sync.RWMutex
+	events map[string]prometheus.Counter
+}
+
+// NewCBMetrics creates circuit breaker metrics backed by the supplied Prometheus
+// registerer. Metrics are registered the first time this function is called with
+// a given registerer. Subsequent calls reuse the existing collectors.
+func NewCBMetrics(reg prometheus.Registerer) CBMetrics {
+	m := &cbMetrics{
+		opens: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "crossplane",
+			Name:      "circuit_breaker_opens_total",
+			Help:      "Number of times the XR circuit breaker transitioned from closed to open.",
+		}, []string{"controller"}),
+		closes: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "crossplane",
+			Name:      "circuit_breaker_closes_total",
+			Help:      "Number of times the XR circuit breaker transitioned from open to closed.",
+		}, []string{"controller"}),
+		events: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "crossplane",
+			Name:      "circuit_breaker_events_total",
+			Help:      "Number of XR watch events handled by the circuit breaker, labelled by outcome.",
+		}, []string{"controller", "result"}),
+	}
+
+	registerCounterVec(reg, &m.opens)
+	registerCounterVec(reg, &m.closes)
+	registerCounterVec(reg, &m.events)
+
+	return m
+}
+
+func registerCounterVec(reg prometheus.Registerer, cv **prometheus.CounterVec) {
+	if reg == nil {
+		return
+	}
+
+	if err := reg.Register(*cv); err != nil {
+		var already prometheus.AlreadyRegisteredError
+		if errors.As(err, &already) {
+			if existing, ok := already.ExistingCollector.(*prometheus.CounterVec); ok {
+				*cv = existing
+				return
+			}
+		}
+		// Avoid crashing the control plane if registration fails; the collector
+		// will continue to function locally even if it's not exported.
+		return
+	}
+}
+
+func (m *cbMetrics) IncOpen(controller string) {
+	if m == nil {
+		return
+	}
+
+	cc := m.getController(controller)
+	cc.opens.Inc()
+}
+
+func (m *cbMetrics) IncClose(controller string) {
+	if m == nil {
+		return
+	}
+
+	cc := m.getController(controller)
+	cc.closes.Inc()
+}
+
+func (m *cbMetrics) IncEvent(controller, result string) {
+	if m == nil {
+		return
+	}
+
+	cc := m.getController(controller)
+
+	cc.mu.RLock()
+	counter, ok := cc.events[result]
+	cc.mu.RUnlock()
+	if ok {
+		counter.Inc()
+		return
+	}
+
+	cc.mu.Lock()
+	counter = cc.events[result]
+	if counter == nil {
+		counter = m.events.WithLabelValues(controller, result)
+		cc.events[result] = counter
+	}
+	cc.mu.Unlock()
+
+	counter.Inc()
+}
+
+func (m *cbMetrics) getController(controller string) *controllerCounters {
```

Comment:
I'm having a hard time following this method.

Why do we need a mutex and a map of counters to track events? Doesn't the Prometheus SDK handle this for us? I'm not seeing why we can't just use `With(labels).Inc()` as we do elsewhere.


https://github.com/crossplane/crossplane/blob/098ee8e6a4827a5d28b879354223d0c3c43acb9c/internal/engine/engine_metrics.go#L77